tls: vec_push: handle async errors rather than throwing on_internal_error

bhalevy opened the pull request:
This series deals with handling of async errors in vec_push _output_pending.

First, `handle_output_error` is called in `do_handshake` before calling `send_alert`, fixing #1180
and on top, the "unhandled error" internal error in vec_push was demoted and is just reported and returned back, since it was determined that it can be hit by normal circumstances when _output_pending fails in background and gnutls is calling vec_push a number of times in a row. This fixes #1181.

debug logging and and error injection test that reproduces the above issue without the fix were also added.

Refs https://github.com/scylladb/scylladb/issues/11252

avikivity commented:
subject lacks a verb

bhalevy commented:
> subject lacks a verb

@avikivity, I fixed the subject

nyh reviewed (approved):
I can't find any error in the code, so Looks Good To Me (TM), but in general the flow of these errors has become so convoluted that I'm not sure I understood everything going on there....
Did you check the send_alert() thing still works? I added it in commit 0e48ba883215508e5a33ae1a093fec901e528cb6 to make sure we generate correct errors in case of handshake failures (not IO failures - real protocol incompatibilities). Looking at the code, I **think** it should still work, but you can test it in Scylla with `test/cql-pytest/run --ssl test_ssl.py` (but I see now that's broken - I'll prepare a patch)

nyh commented:
> you can test it in Scylla with `test/cql-pytest/run --ssl test_ssl.py` (but I see now that's broken - I'll prepare a patch)

Beyond the fact that Scylla's "run --ssl" broke due to a regression in the test suite (which has a trivial fix and I'll send it in a minute), unfortunately there seems to have also been a regression in the error message - now when there is an handshake error because of an unsupported protocol version, instead of a message about "no protocols available" we get "SSL] internal error (_ssl.c:997)". I still don't know why. But anyway, it predates your patch. We can check it later.

avikivity commented on src/net/tls.cc:
```diff
@@ -1113,11 +1113,21 @@ class session : public enable_lw_shared_from_this<session> {
+                        if (f.failed()) {
+                            ex = f.get_exception();
+                        } else {
+                            on_internal_error_noexcept(tlslog, "expected output error after do_handshake error");
```
seastar_logger should be sufficient for something that's expected never to happen.

bhalevy replied:
> handle_output_error always returns an exception, so this case cannot happen

Why?  At this point we know only that `gnutls_handshake` returned an error.
Can we be sure that `_output_pending` (that is returned by `wait_for_output()`) contains an error?

bhalevy replied:
> I wonder if it isn't easier not to even bother unpacking f (which you know is always a failed future) and we can just return f itself? I think it will simplify this code a lot.

Can do, although I think the internal error case is still theoretically possible, if we have a bug.

nyh replied:
> > handle_output_error always returns an exception, so this case cannot happen
> 
> Why? At this point we know only that `gnutls_handshake` returned an error. Can we be sure that `_output_pending` (that is returned by `wait_for_output()`) contains an error?

Maybe I'm missing something, but it seems to me that handle_output_error() unconditionally returns an exception - if wait_for_output() succeeded, it has `return make_exception_future(_errror)`, otherwise it has `std::throw_with_nested()`. Isn't this an exception in any case?

elcallio replied:
Handle output error already does everything this extra continuation does, minus the `on_internal_error_noexcept`, which I don't think is appropriate, because `default` is not same as "IO error".

bhalevy replied:
Looking at it again, I concur.

bhalevy commented on src/net/tls.cc:
```diff
@@ -46,6 +47,8 @@
 
 namespace seastar {
 
+logger tlslog("tls");
```
It can be defined static actually.
And maybe it should be called seastar_tlslog, to make its scope clearer.

elcallio replied:
Should TLS have a log? Why? We don't have loggers for TCP etc? Should not any logger be user-provided for this level of library function?

bhalevy replied:
Maybe we should have finer granularity loggers so their level can be controlled in tests or production.
But this can be discussed orthogonally.
Currently my main concern is getting the fix in.

bhalevy commented on src/net/tls.cc:
```diff
@@ -1381,23 +1381,26 @@ class session : public enable_lw_shared_from_this<session> {
-            on_internal_error(seastar_logger, format("tls: session: vec_push: unhandled pending output error: {}", ex));
-        }
         try {
+            ssize_t n;
```
what else case?
If pending_error is false, we set n, otherwise we rethrow the exception and never get to return n.
I can set to it 0, but I actually want the compiler to warn us if n is returned by some path where it's uninitialized, since I don't want the function to return 0 inadvertently either.

nyh replied:
Maybe I missed something... But if there are really two different paths here, one which sets and returns n, and one which doesn't, can't it be written like the old code where n is declared only in the first path?

bhalevy replied:
the problem is that the good path is broken into two blocks hence n was moved out a scope.

bhalevy commented on src/net/tls.cc:
```diff
@@ -1381,23 +1381,26 @@ class session : public enable_lw_shared_from_this<session> {
-        }
         try {
+            ssize_t n;
+            bool pending_error = _output_pending.failed();
```
I can do this in the `if` condition.
The variable is used below for constructing the error message so we can tell if the error was on _output_pending before the call to `_out.put` or is it a synchronous error from `._out.put()`.  Doesn't make much difference to the user, but could be interesting to a developer later on...

elcallio commented on src/net/tls.cc:
```diff
@@ -1158,14 +1183,19 @@ class session : public enable_lw_shared_from_this<session> {
         }
         return _in.get().then([this](buf_type buf) {
             _eof |= buf.empty();
+            tlslog.trace("wait_for_input: buf size={} eof={}", buf.size(), _eof);
```
I get the desire to printf-debug TLS state machine, but in a real world scenario, since none of this log messages identify the TLS connection in question, they will just be annoying noise once num connections >= 2. I again really don't think a logger is appropriate for this level of support code.

elcallio commented on src/net/tls.cc:
```diff
@@ -1355,27 +1386,38 @@ class session : public enable_lw_shared_from_this<session> {
+                _output_pending = _output_pending.then([] {
+                    return make_exception_future<>(std::runtime_error("injected vec_push error"));
+                });
             }
```
Again, can accomplish this by just writing the test to run on a pipe and inject the errors there.

elcallio commented:
See `test_close_timout` for example of using a pipe-subclass to generate wire errors for test.

bhalevy commented on src/net/tls.cc:
```diff
@@ -38,14 +38,21 @@
 
+#ifdef SEASTAR_DEBUG
+bool tls_inject_handshake_error = false;
+#endif
```
I tried something like that first, but to reproduce the problem I had to to set the error only when `_out.put()` returned a non-ready future.  I'm not sure if that's doable with error-injecting pipe implementation.
I suggest to get the fix first and then wok on the test as a follow up since without it scylla tends to crash quite frequently in upgrade tests (https://github.com/scylladb/scylladb/issues/11252)

bhalevy commented:
In v2 (7cff49b5):
- dropped the tlslog patch
- dropped the error injection test
- simplified handle_output_error after gnutls_handshake failure, and before send_alert
- clarified vec_push error handling implementation

elcallio commented on src/net/tls.cc:
```diff
@@ -1358,23 +1360,25 @@ class session : public enable_lw_shared_from_this<session> {
                 // by the catch handlers below
-                std::rethrow_exception(_output_pending.get_exception());
+                auto level = output_pending_failed ? log_level::warn : log_level::debug;
+                seastar_logger.log(level, "tls: vec_push: {}output failed: {}", output_pending_failed ? "pending " : "", ex);
```
Why the logging? To try to find a deeper issue in test runs? If so should be marked as temporary. If not it should not be here. We've established that reaching vec_push with _output_pending in error state (or any state except ready waited for) is an error, i.e. some oversight in the state machine of our tls.

elcallio replied:
At the very least this logging should only be done in the actual real error case, i.e. we came in with a pending error.

elcallio reviewed (changes requested):
I don't think the debug-log on "normal" error is warranted.

bhalevy commented:
> I don't think the debug-log on "normal" error is warranted.

yeah, leftover from the previous series.
Will remove it.

bhalevy commented:
In v3 (8cfb018f):
- renamed `output_pending_failed` variable to `prev_output_pending_failed` to make it even clearer.
- warn only if prev_output_pending_failed happened, otherwise log nothing.

avikivity commented on src/net/tls.cc:
```diff
@@ -1360,23 +1360,27 @@ class session : public enable_lw_shared_from_this<session> {
-                std::rethrow_exception(_output_pending.get_exception());
+                auto ex = _output_pending.get_exception();
+                if (prev_output_pending_failed) {
+                    seastar_logger.warn("tls: vec_push: pending output failed: {}", ex);
```
Isn't this user-triggerable? A user can connect to a server and spam the logs this way.

Since we rethrow the exception, the caller can decide whether to log the error or not.

bhalevy replied:
How?
It's printed only in the rare case of an asynchromous error that isn't caught by wait_for_output.  It's actually quite hard to reproduce, even with error injection.

avikivity replied:
I don't see that it merits a warning. What can the user do about it? report it to the application author?

I think it's okay to collapse two errors into one (as I think the code is doing) without additional noise in the logs. Swallowing errors is bad, but this isn't the case here.

bhalevy replied:
Ok, the audience for this error is us indeed, not the user, who can't do anything about it.

avikivity commented:
@elcallio let's plan to write an error injection test using loopback_socket that can inject errors at any byte boundary, this looks like a nest of bugs.

bhalevy commented:
> I can't reproduce the failure seen in CI. I don't know if it's due to existing flakiness or is it caused by the changes in this PR.

Same failure seen in https://app.circleci.com/pipelines/github/scylladb/seastar/1384/workflows/34b5e768-acb6-4503-a49c-54977ea74c2f/jobs/3517/parallel-runs/0/steps/0-105

For #1178 , proving it's a preexisting ussue, possibly trigerred by the test environment.

bhalevy commented:
@avikivity is there a way to rerun the CI tests?
Although given that fact the same failures were seen independently on another PR maybe this wouldn't be needed.

benipeled commented:
> I can't reproduce the failure seen in CI. I don't know if it's due to existing flakiness or is it caused by the changes in this PR.

@bhalevy seems like https://github.com/scylladb/seastar/issues/1127 which AFAIK marked as circle-ci infra issue @xemul ?

In any case, you can rerun it from the CI page https://app.circleci.com/pipelines/github/scylladb/seastar/1389

bhalevy commented:
> > I can't reproduce the failure seen in CI. I don't know if it's due to existing flakiness or is it caused by the changes in this PR.
> 
> @bhalevy seems like #1127 which AFAIK marked as circle-ci infra issue @xemul ?
> 
> In any case, you can rerun it from the CI page https://app.circleci.com/pipelines/github/scylladb/seastar/1389

Thanks. Just it says that I need write permissions to trigger rebuild.

avikivity commented:
You can always trigger CI on your own branch by linking your github account to circleci.

bhalevy commented:
ping. I couldn't rerun the CI job.
I don't think it's necessary since the failure is clearly due to a preexisting issue.
Maintainers, if you think it's crucial please run the CI job again.

Cc @slivne

bhalevy commented:
f98ce73e: rebased to re-trigger CI

bhalevy commented:
CI failure seems unrelated to this change.
I opened https://github.com/scylladb/seastar/issues/1190 to capture it
